File: src/try.py
from math import ceil

def enable_connections(graph, l, b, node):
	# left
	if node % l != 1:
		graph[node - 1][node - 2] = 1
		graph[node - 2][node - 1] = 1
	# right
	if node % l != 0:
		graph[node - 1][node] = 1
		graph[node][node - 1] = 1
	# top
	if node <= l * (b-1):
		graph[node - 1][node + l - 1] = 1 
		graph[node + l - 1][node - 1] = 1 
	# down
	if node > l:
		graph[node - 1][node - l - 1] = 1
		graph[node - l - 1][node - 1] = 1

def disable_connections(graph, node):
	for i in range(len(graph[node - 1])):
		graph[node - 1][i] = 0
		graph[i][node - 1] = 0

def build_graph(dest):
	l = ceil(dest[0])
	b = ceil(dest[1])
	n = l * b
	graph = [ [ 0 for i in range(n) ] for i in range(n) ]
	for i in range(1, n+1):
		enable_connections(graph, l, b, i)
	return l, b, graph

def find_adjacent_nodes(graph, node, visited):
	def _is_visited(_v, _n):
		_r = False
		for _tn in _v.keys():
			if _tn == _n:
				_r = True
				break
		return _r
	result = []
	for temp_node, connection in enumerate(graph[node - 1]):
		if connection == 1 and not _is_visited(visited, temp_node + 1):
			result.append(temp_node + 1)
	return result

def find_path(graph, start_node, end_node):
	path = []
	
	visited = {start_node:None}
	temp = find_adjacent_nodes(graph, start_node, visited)
	for t in temp:
		visited[t] = start_node
	
	queue = []
	queue.extend( temp )
	
	prev_node = start_node
	while len(queue) != 0:
		curr_node = queue.pop()
		# visited[curr_node] = prev_node
		temp = find_adjacent_nodes(graph, curr_node, visited)
		for t in temp:
			visited[t] = curr_node
		# print('curr_node:', curr_node)
		# print('adjacent_nodes: ', temp)
		# print('visited: ', visited)
		# print('queue: ', queue)
		# print()
		if curr_node == end_node: break
		queue.extend( temp )
		prev_node = curr_node

	# print('\n\n\npath finding done\n\n\n')

	if end_node not in visited.keys():
		print('Cannot find path from', start_node, 'to', end_node)
	else:
		# print('path possible!')
		#print(visited)
		# for i in sorted(visited.items()):
			# print(i)
		temp = end_node
		path.append(temp)
		while temp != start_node:
			# print(path)
			# print('from', temp, 'to', visited[temp])
			path.append(visited[temp])
			temp = visited[temp]
		path.reverse()
	
	return path

def print_graph(graph):
	print('   ', end='')
	for i in range(len(graph)):
		print(i+1, ' ', end='')
	print()
	for i, r in enumerate(graph):
		print(i + 1, r)

# if __name__ == "__main__":
# 	l, b, graph = build_graph((1.3, 2.4))
# 	print_graph(graph)
# 	disable_connections(graph, 3)
# 	print_graph(graph)

# 	path = find_path(graph, 1, 6)

# 	print('path', path)

# if __name__ == "__main__":
# 	l, b, graph = build_graph((2.3,2.3))
# 	print_graph(graph)
# 	disable_connections(graph, 5)
# 	disable_connections(graph, 7)
# 	# disable_connections(graph, l, b, 1)
# 	print_graph(graph)

# 	path = find_path(graph, 4, 9)

# 	print('path', path)

if __name__ == "__main__":
	l, b, graph = build_graph((5.6,5.6))
	# print_graph(graph)
	for i in [13,14,15,26,27,28,29,30]:
		disable_connections(graph, i)
	# disable_connections(graph, l, b, 1)
	# print_graph(graph)

	path = find_path(graph, 6, 36)

	print('path', path)

File: src/try.cpp
#include <vector>
#include <map>
#include <queue>
#include <cmath>
#include <algorithm>
#include <iostream>

std::vector<std::vector<int> > graph;

void enable_connections(std::vector<std::vector<int> > &graph, int l, int b, int node) {
	// left
	if(node % l != 1) {
		graph[node - 1][node - 2] = 1;
		graph[node - 2][node - 1] = 1;
	}
	// right
	if(node % l != 0) {
		graph[node - 1][node] = 1;
		graph[node][node - 1] = 1;
	}
	// top
	if(node <= l * (b-1)) {
		graph[node - 1][node + l - 1] = 1;
		graph[node + l - 1][node - 1] = 1;
	}
	// down
	if(node > l) {
		graph[node - 1][node - l - 1] = 1;
		graph[node - l - 1][node - 1] = 1;
	}
}

void disable_connections(std::vector<std::vector<int> > &graph, int node) {
	for(int i = 0; i < graph[node - 1].size(); i++) {
		graph[node - 1][i] = 0;
		graph[i][node - 1] = 0;
	}
}

void build_graph(float x, float y, int &l, int &b) {
	l = std::ceil(x);
	b = std::ceil(y);
	
	int n = l * b;

	for(int i = 0; i < n; i++) {
		graph.push_back(std::vector<int>(n));
	}

	for(int i = 1; i <= n; i++) {
		enable_connections(graph, l, b, i);
	}
}

bool _is_visited(const std::map<int, int> &visited, int node) {
	return visited.find(node) != visited.end();
}

void find_adjacent_nodes(const std::vector<std::vector<int> > &graph, const int node, const std::map<int, int> &visited, std::vector<int> &result) {
	int n = graph[node - 1].size();
	for(int i = 0; i < n; i++) {
		if(graph[node - 1][i] == 1 && !_is_visited(visited, i + 1)) {
			result.push_back(i + 1);
		}
	}
}

void find_path(const std::vector<std::vector<int> > &graph, const int start_node, const int end_node, std::vector<int> &path) {
	std::map<int, int> visited;

	visited[start_node] = -1;

	std::vector<int> temp;

	find_adjacent_nodes(graph, start_node, visited, temp);

	std::queue<int> queue;
	
	for(auto e : temp) {
		visited[e] = start_node;
		queue.push(e);
	}
	temp.clear();

	int prev_node = start_node;

	while(queue.size() != 0) {
		int curr_node = queue.front();
		queue.pop();

		find_adjacent_nodes(graph, curr_node, visited, temp);

		for(int e : temp) {
			visited[e] = curr_node;
			queue.push(e);
		}
		temp.clear();
		
		if(curr_node == end_node) {
			break;
		}

		prev_node = curr_node;
	}

	if(visited.find(end_node) != visited.end()) {
		std::cout << "Wohhoo\n";
		int temp_node = end_node;
		path.push_back(temp_node);
		while(temp_node != start_node) {
			path.push_back(visited[temp_node]);
			temp_node = visited[temp_node];
		}
		std::reverse(path.begin(), path.end());
	}
	else {
		std::cout << "Shaata\n";
	}
}

int main() {
	int l;
	int b;
	build_graph(5.6, 5.6, l, b);

	for(auto e1 : graph) {
		for(auto e2 : e1) {
			std::cout << e2 << " ";
		}
		std::cout << std::endl;
	}
	std::cout << std::endl;

	int temp[] = {13,14,15,26,27,28,29,30};//,    4, 5,6};
	for(auto e1 : temp)
		disable_connections(graph, e1);

	for(auto e1 : graph) {
		for(auto e2 : e1) {
			std::cout << e2 << " ";
		}
		std::cout << std::endl;
	}
	std::cout << std::endl;

	std::vector<int> path;
	find_path(graph, 1, 36, path);

	for(auto e : path) {
		std::cout << e << " -> ";
	}
	std::cout << "\n";

	return 0;
}

/*

	1 -> 2 -> 3 -> 9 -> 10 -> 16 -> 22 -> 21 -> 20 -> 19 -> 25 -> 31 -> 32 -> 33 -> 34 -> 35 -> 36 ->



*/